File: app/src/main/java/wangdaye/com/geometricweather/location/services/BaiduIPLocationService.java
package wangdaye.com.geometricweather.location.services;

import android.content.Context;

import android.text.TextUtils;
import androidx.annotation.NonNull;

import javax.inject.Inject;

import io.reactivex.rxjava3.disposables.CompositeDisposable;
import wangdaye.com.geometricweather.common.rxjava.BaseObserver;
import wangdaye.com.geometricweather.common.rxjava.ObserverContainer;
import wangdaye.com.geometricweather.common.rxjava.SchedulerTransformer;
import wangdaye.com.geometricweather.location.apis.BaiduIPLocationApi;
import wangdaye.com.geometricweather.location.json.BaiduIPLocationResult;
import wangdaye.com.geometricweather.settings.SettingsManager;

public class BaiduIPLocationService extends LocationService {

    private final BaiduIPLocationApi mApi;
    private final CompositeDisposable compositeDisposable;

    @Inject
    public BaiduIPLocationService(BaiduIPLocationApi api,
                                  CompositeDisposable disposable) {
        mApi = api;
        compositeDisposable = disposable;
    }

    @Override
    public void requestLocation(Context context, @NonNull LocationCallback callback) {
        mApi.getLocation(SettingsManager.getInstance(context).getProviderBaiduIpLocationAk(), "gcj02")
                .compose(SchedulerTransformer.create())
                .subscribe(new ObserverContainer<>(compositeDisposable, new BaseObserver<BaiduIPLocationResult>() {
                    @Override
                    public void onSucceed(BaiduIPLocationResult baiduIPLocationResult) {
                        if (baiduIPLocationResult.getContent() == null
                                || baiduIPLocationResult.getContent().getPoint() == null
                                || TextUtils.isEmpty(baiduIPLocationResult.getContent().getPoint().getY())
                                || TextUtils.isEmpty(baiduIPLocationResult.getContent().getPoint().getX())) {
                            callback.onCompleted(null);
                        } else {
                            try {
                                Result result = new Result(
                                        Float.parseFloat(baiduIPLocationResult.getContent().getPoint().getY()),
                                        Float.parseFloat(baiduIPLocationResult.getContent().getPoint().getX())
                                );
                                callback.onCompleted(result);
                            } catch (Exception ignore) {
                                callback.onCompleted(null);
                            }
                        }
                    }

                    @Override
                    public void onFailed() {
                        callback.onCompleted(null);
                    }
                }));
    }

    @Override
    public void cancel() {
        compositeDisposable.clear();
    }

    @Override
    public boolean hasPermissions(Context context) {
        return true;
    }

    @Override
    public String[] getPermissions() {
        return new String[0];
    }
}

File: app/src/main/java/org/breezyweather/sources/baiduip/BaiduIPLocationService.kt
package org.breezyweather.sources.baiduip

import android.content.Context
import io.reactivex.rxjava3.core.Observable
import org.breezyweather.common.exceptions.ApiKeyMissingException
import org.breezyweather.common.exceptions.LocationException
import org.breezyweather.common.rxjava.SchedulerTransformer
import org.breezyweather.common.source.HttpSource
import org.breezyweather.common.source.LocationPositionWrapper
import org.breezyweather.common.source.LocationSource
import org.breezyweather.settings.SettingsManager
import retrofit2.Retrofit
import javax.inject.Inject

class BaiduIPLocationService @Inject constructor(
    client: Retrofit.Builder
) : HttpSource(), LocationSource {

    override val id = "baidu_ip"
    override val name = "百度IP定位"
    override val privacyPolicyUrl = "https://lbs.baidu.com/index.php?title=openprivacy"

    private val mApi by lazy {
        client
            .baseUrl(BAIDU_IP_LOCATION_BASE_URL)
            .build()
            .create(BaiduIPLocationApi::class.java)
    }

    override fun requestLocation(context: Context): Observable<LocationPositionWrapper> {
        val apiKey = SettingsManager.getInstance(context).providerBaiduIpLocationAk
        if (apiKey.isEmpty()) {
            return Observable.error(ApiKeyMissingException())
        }
        return mApi.getLocation(apiKey, "gcj02")
            .compose(SchedulerTransformer.create())
            .map { t ->
                if (t.status != 0) {
                    // 0 = OK
                    // 1 = IP not supported (outside China)
                    // Don’t know about other cases, doing != 0 for safety
                    throw LocationException()
                }
                if (t.content?.point == null
                    || t.content.point.y.isNullOrEmpty()
                    || t.content.point.x.isNullOrEmpty()
                ) {
                    throw LocationException()
                } else {
                    try {
                        LocationPositionWrapper(
                            t.content.point.y.toFloat(),
                            t.content.point.x.toFloat()
                        )
                    } catch (ignore: Exception) {
                        throw LocationException()
                    }
                }
            }
    }

    override fun hasPermissions(context: Context) = true

    override val permissions: Array<String> = emptyArray()

    companion object {
        private const val BAIDU_IP_LOCATION_BASE_URL = "https://api.map.baidu.com/"
    }
}
